.ci/azure-pipelines/azure-pipelines.yaml

```diff
@@ -14,29 +16,114 @@ stages:
     displayName: Formatting
     jobs:
       - template: formatting.yaml
-  - stage: build_ubuntu
-    displayName: Build Ubuntu
+
+  - stage: build_gcc
+    displayName: Build GCC
     dependsOn: formatting
     jobs:
-      - template: build-ubuntu.yaml
-  - stage: build_osx
-    displayName: Build macOS
+      - job: ubuntu
+        displayName: Ubuntu
+        pool:
+          vmImage: 'Ubuntu 16.04'
+        strategy:
+          matrix:
+            16.04 GCC:
+              CONTAINER: env1604
+              CC: gcc
+              CXX: g++
+              CMAKE_ARGS: '-DPCL_WARNINGS_ARE_ERRORS=ON'
+            20.04 GCC:
+              CONTAINER: env2004
+              CC: gcc
+              CXX: g++
+              CMAKE_ARGS: ''
+        container: $[ variables['CONTAINER'] ]
+        timeoutInMinutes: 0
+        variables:
+          BUILD_DIR: '$(Agent.BuildDirectory)/build'
+          CMAKE_CXX_FLAGS: '-Wall -Wextra'
+        steps:
+          - template: build/build-ubuntu.yaml
+
+  - stage: build_clang
+    displayName: Build Clang
     dependsOn: formatting
     jobs:
-      - template: build-macos.yaml
-  - stage: build_windows
-    displayName: Build Windows
+      - job: osx
+        displayName: macOS
+        pool:
+          vmImage: '$(VMIMAGE)'
+        strategy:
+          matrix:
+            Catalina 10.15:
+              VMIMAGE: 'macOS-10.15'
+              OSX_VERSION: '10.15'
+            Mojave 10.14:
+              VMIMAGE: 'macOS-10.14'
+              OSX_VERSION: '10.14'
+        timeoutInMinutes: 0
+        variables:
+          BUILD_DIR: '$(Agent.WorkFolder)/build'
+          GOOGLE_TEST_DIR: '$(Agent.WorkFolder)/googletest'
+          CMAKE_CXX_FLAGS: '-Wall -Wextra -Werror -Wabi'
+        steps:
+          - template: build/build-macos.yaml
+      - job: ubuntu
+        displayName: Ubuntu
+        dependsOn: osx
+        condition: succeededOrFailed()
+        pool:
+          vmImage: 'Ubuntu 16.04'
+        strategy:
```

Comment:
One or two lines explaining the reasoning behind the way the tests are run would be helpful. Eg: allows Mac, Windows and Ubuntu to be run in parallel. Placement of Ubuntu-clang job reduces need of parallel CI jobs and doesn't impact total time per run under ideal conditions.